src/track/track.cpp

```diff
@@ -337,13 +337,27 @@ bool Track::setBeatsWhileLocked(mixxx::BeatsPointer pBeats) {
     return true;
 }
 
-void Track::setBeatsMarkDirtyAndUnlock(QMutexLocker* pLock, mixxx::BeatsPointer pBeats) {
+void Track::setBeatsMarkDirtyAndUnlock(
+        QMutexLocker* pLock,
+        mixxx::BeatsPointer pBeats,
+        bool bpmLocked) {
     DEBUG_ASSERT(pLock);
+
+    if (m_pBeats && m_record.getBpmLocked()) {
```

Comment:
This doesn't work as expected if you pass `false` but the bpm is still locked.

This internal function should simply set both members and set the dirty flag accordingly. No implicit logic.

The bpm locked flag should be checked at the public API and if an update has been rejected this must be indicated at least by the return value. The function name should clearly indicate that the bpmLocked flags is evaluated and the request might be rejected, i.e. `bool setBpmIfNotLocked()` or `bool setBeatsIfNotLocked()`. Otherwise you would expect that `set` is executed unconditionally.

Reply:
For the public API even using `[[nodiscard]]` could make sense to prevent silently ignoring the return value at the call site.

Reply:
Now it always applied the passed bpm lock flag, if the beats where set successful. I hope if it is clear form the new parameter name.